app/controllers/admin/patient_imports_controller.rb

```diff
@@ -0,0 +1,25 @@
+class Admin::PatientImportsController < AdminController
+  include FileUploadable
+
+  def new
+    authorize { current_admin.power_user? }
+  end
+
+  def create
+    authorize { current_admin.power_user? }
+
+    facility = Facility.find(params[:facility_id])
+
+    data = read_xlsx_or_csv_file(params[:patient_import_file])
+    params = PatientImport::SpreadsheetTransformer.call(data, facility: facility)
+    errors = PatientImport::Validator.new(params).errors
+
+    if errors.values.flatten.any?
+      @errors = errors
+      render :new
+    else
+      results = PatientImport::Importer.call(params: params, facility: facility, admin: current_admin)
```

Comment:
Consider moving to a background job, and email when done? This is going to block a web request and could very easily time out for any large spreadsheet.

Reply:
That sounds like a good idea. 150 patients was taking over 30 seconds to process.

I will also disable retries on the job, since the import is not designed to be idempotent.

Reply:
👍

Reply:
This point is now moot, since every patient gets their own job. 👍